doc/content/integrations/webhooks/cloud-integrations/ubidots/ubidots-setup/_index.md

```diff
@@ -0,0 +1,99 @@
+---
+title: "Ubidots Setup"
+description: ""
+weight: 1
+---
+
+Create a function by following these instructions to prepare Ubidots setup for the integration with {{% tts %}}. This function will handle the conversion of incoming messages from JSON format to the one that is compatible with Ubidots.
+
+<!--more-->
+
+Log in to your Ubidots account and find **Devices** tab in the upper part of your dashboard. In its drop-down list, choose **Functions**.
+
+When redirected to **UbiFunctions** page, create a new function with **Create Function** button.
+
+On the left, give a **Name** to your function, select **POST** method and choose **Python 3.6** for **Runtime**.
+
+In function window, paste this code:
+
+```
+import requests
+import json
+import time
+
+BASE_URL = "https://industrial.api.ubidots.com"
+TOKEN = "····" # Enter a token here
+
+def main(args):
+    # Prining args from TTI
+    print(f'[INFO] Args from TTI:\n {args}')
+
+    # Parsing data
+    payload = parse_tti_data(args)
+    dev_label = tti_dev_eui(args)
+    print(f'[INFO] Parsed data:\n {payload}')
+    print(f'[INFO] TTI Dev_EUI data:\n {dev_label}')
+
+    # Posting to Ubidots
+    req = update_device(dev_label, payload, TOKEN)
+    print(f'[INFO] Request to Ubidots Status code: {req.status_code}')
+    print(f'[INFO] Request ti Ubidots JSON:\n {req.json()}')
+
+    return {
+        'status_code': req.status_code,
+        'response_json': req.json()
+    }
+
+
+def parse_tti_data(data):
+    return data['uplink_message']['decoded_payload']
+
+
+def tti_dev_eui(data):
+    return data['end_device_ids']['device_id']
+
+
+def update_device(device, payload, token):
+    """
+    Updates device with payload
+    """
+    url = "{}/api/v1.6/devices/{}".format(BASE_URL, device)
+    headers = {"X-Auth-Token": token, "Content-Type": "application/json"}
+    req = create_request(url, headers, attempts=5, request_type="post", data=payload)
+    return req
+
+
+def create_request(url, headers, attempts, request_type, data=None):
+    """
+    Function to make a request to the server
+    """
+    request_func = getattr(requests, request_type)
+    kwargs = {"url": url, "headers": headers}
+    if request_type == "post" or request_type == "patch":
+        kwargs["json"] = data
+    try:
+        req = request_func(**kwargs)
+        status_code = req.status_code
+        time.sleep(1)
+        while status_code >= 400 and attempts < 5:
+            req = request_func(**kwargs)
+            status_code = req.status_code
+            attempts += 1
+            time.sleep(1)
+        return req
+    except Exception as e:
+        print("[ERROR] There was an error with the request, details:")
+        print(e)
+        return None
+```
+Since this function needs your token to be entered on the sixth line, you can find and copy it from **Tokens** if you click on your avatar in the upper right corner and select **API Credentials**. 
```

Comment:
Is this a Ubidots token or tts? Why does the Ubidots function need a token if its running inside Ubidots?

Reply:
It's a Ubidots token and is used because the UbiFunction engine acts like a mediator - it takes JSON from TTS, then parses it and makes an API request to post data to Ubidots. The HTTP POST from TTS webhook is addressed on the UbiFunction, while the UbiFunction makes an actual POST request to Ubidots after processing the message. I was following their example: https://help.ubidots.com/en/articles/3975015-integrate-your-the-things-industries-data-with-ubidots-using-ubifunctions